### DISCUSSION TITLE
The router API should allow its backend to be configurable

### DISCUSSION THREAD

**Discussion by dead-claudia** [MithrilJS/mithril.js#2942] [Ideas]
It should be possible to configure how URLs are routed, so people aren't having to fork it for use cases like [this](https://gitter.im/mithriljs/mithril.js?at=5a2ed2b5c65707ba2b3cc0a4). For comparison, it'd give us the ability to provide a hook like [React Router's `history` attribute](https://reacttraining.com/react-router/core/api/history), without much work on our part.

(It's come up on Gitter more than once. Marking for v2 because it's a pretty trivial change on our part.)

Here's what I propose:

- Instead of accepting a [mock `$window` directly](https://github.com/MithrilJS/mithril.js/blob/next/router/router.js#L6) for the core router, accept a backend, and default to one that wraps `$window`. Same goes for [the API frontend](https://github.com/MithrilJS/mithril.js/blob/next/api/router.js#L7).
- For the backend, require the following methods and use them in lieu of the given blocks:
    - [`backend.normalize(fragment)`](https://github.com/MithrilJS/mithril.js/blob/next/router/router.js#L10-L14) - Get and normalize the specified fragment.
    - [`backend.setPath(prefix, path, state, title, replace)`](https://github.com/MithrilJS/mithril.js/blob/next/router/router.js#L69-L76) - Set the path given certain options
    - [`backend.initialState()`](https://github.com/MithrilJS/mithril.js/blob/next/router/router.js#L84) - Get the initial state to potentially assign.
    - [`backend.registerResolver(resolver)`](https://github.com/MithrilJS/mithril.js/blob/next/router/router.js#L107-L108) - Register the router's `resolveRoutes` function.

Basically, I don't want to require every user that needs to override the underlying routing mechanism to redo the entire router, but instead just require them to implement a few methods.

**Comment by dead-claudia** [comment]
If I find time, I'll implement this. It's pretty straightforward.
